db/kvstorewithcache.go

```diff
@@ -0,0 +1,125 @@
+package db
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/iotexproject/go-pkgs/cache"
+	"github.com/iotexproject/iotex-core/db/batch"
+)
+
+// kvStoreWithCache is an implementation of KVStore, wrapping kvstore with LRU caches of latest states
+type kvStoreWithCache struct {
+	store       KVStore
+	stateCaches map[string]*cache.ThreadSafeLruCache // map having lru cache to store current states for speed up
+	cacheSize   int
+}
+
+// NewKvStoreWithCache wraps kvstore with stateCaches
+func NewKvStoreWithCache(kvstore KVStore, cacheSize int) KVStore {
+	return &kvStoreWithCache{
+		store:       kvstore,
+		stateCaches: make(map[string]*cache.ThreadSafeLruCache),
+		cacheSize:   cacheSize,
+	}
+}
+
+// Start starts the kvStoreWithCache
+func (kvc *kvStoreWithCache) Start(ctx context.Context) error {
+	return kvc.store.Start(ctx)
+}
+
+// Stop stops the kvStoreWithCache
+func (kvc *kvStoreWithCache) Stop(ctx context.Context) error {
+	for _, sc := range kvc.stateCaches {
+		sc.Clear()
+	}
+	return kvc.store.Stop(ctx)
+}
+
+// Put inserts a <key, value> record into stateCaches and kvstore
+func (kvc *kvStoreWithCache) Put(namespace string, key, value []byte) (err error) {
+	kvc.putStateCaches(namespace, key, value)
+	return kvc.store.Put(namespace, key, value)
+}
+
+// Get retrieves a <key, value> record from stateCaches, and if not exists, retrieves from kvstore
+func (kvc *kvStoreWithCache) Get(namespace string, key []byte) ([]byte, error) {
+	cachedData, err := kvc.getStateCaches(namespace, key)
+	if err != nil {
+		return nil, err
+	}
+	if cachedData != nil {
+		return cachedData, nil
+	}
+	return kvc.store.Get(namespace, key)
+}
```

Comment:
we need add read result into the cache. and that is a key reason for this cache which is increasing state read speed